napari/utils/_injection/__init__.py

```diff
@@ -0,0 +1,125 @@
+from functools import wraps
+from inspect import isgeneratorfunction, signature
+from typing import Any, Callable, Dict, Optional, TypeVar
+
+from typing_extensions import get_type_hints
+
+from ... import components, layers, viewer
+from ._accessors import get_accessor, set_accessor
+from ._processors import get_processor, set_processor
+
+T = TypeVar("T")
+_NULL = object()
+
+
+__all__ = [
+    'get_accessor',
+    'get_processor',
+    'inject_napari_dependencies',
+    'napari_type_hints',
+    'set_accessor',
+    'set_processor',
+]
+
+
+def napari_type_hints(obj: Any) -> Dict[str, Any]:
+    """variant of get_type_hints with napari namespace awareness."""
+    import napari
+
+    return get_type_hints(
+        obj,
+        {
+            'napari': napari,
+            **viewer.__dict__,
+            **layers.__dict__,
+            **components.__dict__,
+        },
+    )
+
+
+def inject_napari_dependencies(func: Callable[..., T]) -> Callable[..., T]:
```

Comment:
If I understood correctly, this decorator does more than just dependency injection as described in the docstring. It is also causes napari to process/consume the return value in some way (e.g. add a layer to the layerlist). Is that right? If so, I think it should be documented here.

After reading the rest of the PR, it might be better to rewrite the docstring in terms of accessors and processors? I think this function with find accessors to inject parameters and will also process return values. I wonder if we might want to split this into two functions (one for accessors/injection and one for processors), but that doesn't have to be in this PR.

Reply:
> It is also causes napari to process/consume the return value in some way (e.g. add a layer to the layerlist). Is that right? If so, I think it should be documented here.

yeah, it will after this PR.  good catch.

> I wonder if we might want to split this into two functions (one for accessors/injection and one for processors), but that doesn't have to be in this PR.

I really like this.  You should be able to inject dependencies without having the output be processed (vice versa, maybe, but less important probably, since you can still override params at call time).  But, it would still be nice to have a single decorator that does both as a convenience, so do you prefer:
1. have three decorators: one for param injection, one for processors, and one for both
2. have two decorators: one for param injection with an optional `process_output: bool` argument, and one just for processors
3. other?

Reply:
Given we know we want the single decorator for both, let's do that first as in this PR, then define two separate ones when needed. I dislike a boolean argument here because it adds branching behavior to the function when we could easily split the function into two.

Reply:
>I dislike a boolean argument here because it adds branching behavior to the function when we could easily split the function into two.

the bool True would simply call the second decorator as a convenience, but they would still be split into two:

```python
def inject_processor(func):
    @wraps(func)
    def _wrapped(*args, **kwargs):
        result = func(*args, **kwargs)
        process_output(result)
        return result
    return _wrapped

def inject_providers(func, process_output: bool=True):
    @wraps(func)
    def _wrapped(*args, **kwargs):
        # inject deps into args kwargs
         return func(*args, **kwargs)
    
    if process_output:
            return inject_processor(_wrapped)
    return _wrapped
```

Reply:
I mostly just have a preference for composition over branching/indentation in the spirit of one function should handle one behavior. Doesn't matter too much here as the functions here are simple and the branching logic is simple, but in general I try to avoid boolean arguments.

```python
def process_output(func):
    @wraps(func)
    def _wrapped(*args, **kwargs):
        result = func(*args, **kwargs)
        process(result)
        return result
    return _wrapped

def inject_providers(func):
    @wraps(func)
    def _wrapped(*args, **kwargs):
        # inject deps into args kwargs
        return func(*args, **kwargs)
    return _wrapped

def inject_providers_and_processor(func):
    return inject_processor(inject_providers(func))
```

Or if I understood correctly, the ordering of wrapping doesn't matter here, so maybe just add both decorators when needed?

```python
@inject_processor
@inject_providers
def foo(...):
    ...
```

Anyway, completely up to you which way to go. I'm approving regardless.

Reply:
Actually, one thing I like about the separate decorators is that I don't love `@inject_processor`, as I don't think of what that's doing as injection. It's more like a post-processor or a done callback for a future - having separate decorators means you could call that something else unrelated to injection (e.g. `@process_output`).

Reply:
very much agree all around.  The naming needs work (and it's not processor "injection").  Let's do in a followup.  Also not opposed to requiring the two decorators... but still holding out hope for a single decorator if one wants both